### PR TITLE
Fix #30091 throwing exceptions

### LINKED ISSUE #30091: AbstractRemoveUnusedVariableCodeFixProvider exception
**Version Used**: 
2.9.0-beta4-63001-02

See work item 685771

**Steps to Reproduce**:

1. Not sure exactly, came in via twitter - https://twitter.com/dannyackerman/status/1041776550000173056

**Expected Behavior**:
The code fix should work.

**Actual Behavior**:
```
ERROR [2018-09-16 20:41:23Z]: Roslyn error: Extension_Exception CSharpRemoveUnusedVariableCodeFixProvider : Sequence contains more than one element
  at System.Linq.Enumerable.Single[TSource] (System.Collections.Generic.IEnumerable`1[T] source) [0x00071] in /Users/builder/jenkins/workspace/build-package-osx-mono/2018-02/external/bockbuild/builds/mono-x64/external/corefx/src/System.Linq/src/System/Linq/Single.cs:46 
  at System.Linq.ImmutableArrayExtensions.Single[T] (System.Collections.Immutable.ImmutableArray`1[T] immutableArray) [0x00007] in <36486b016d234fca8cd67892bf29c7b5>:0 
  at Microsoft.CodeAnalysis.RemoveUnusedVariable.AbstractRemoveUnusedVariableCodeFixProvider`3[TLocalDeclarationStatement,TVariableDeclarator,TVariableDeclaration].RegisterCodeFixesAsync (Microsoft.CodeAnalysis.CodeFixes.CodeFixContext context) [0x00014] in /_/src/Features/Core/Portable/RemoveUnusedVariable/AbstractRemoveUnusedVariableCodeFixProvider.cs:35 
  at Microsoft.CodeAnalysis.CodeFixes.CodeFixService+<GetCodeFixesAsync>d__15.MoveNext () [0x0006f] in /_/src/EditorFeatures/Core/Implementation/CodeFixes/CodeFixService.cs:293 
```

I'll submit a PR, just wanted a bug logged.

### PR DESCRIPTION
**Pull request by Therzok**
The general pattern when registering code fixes seems to be
either passing diagnostics[0] or diagnostics.First().

RemoveUnusedVariable seems to be the only outlier in this case,
doing .Single().

Having multiple diagnostics on a span would cause the code fix
to throw an exception

<details><summary>Ask Mode template not completed</summary>

What does the customer do to get into this situation, and why do we think this
is common enough to address for this release.  (Granted, sometimes this will be
obvious "Open project, VS crashes" but in general, I need to understand how
common a scenario is)

(either VSO or GitHub links)

Also, why we think they are insufficient for RC vs. RC2, RC3, or RTW

This is generally a measure our how central the affected code is to adjacent
scenarios and thus how likely your fix is to destabilize a broader area of code

(with a brief justification for that assessment (e.g. "Low perf impact because no extra allocations/no complexity changes" vs. "Low")

How did we miss it?  What tests are we adding to guard against it in the future?

(E.g. customer reported it vs. ad hoc testing)

If this is a new non-compiler feature or a significant improvement to an existing feature, update https://github.com/dotnet/roslyn/wiki/Manual-Testing once you know which release it is targeting.

</details>

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@mavasani At one point you explained what was up here and what the right pattern should be.  Can you write it up again here.  I'd like to do an audit of our analyzers to make sure they're doing the right thing.

For example:
1. should you process just the first diagnostic?  Or should you process them all.
2. when registering fixes, you can pass an array of diagnostics.  Should you pass just the first diagnostics?  or should you pass all the context diagnostics?

**Comment by CyrusNajmabadi** [comment]
> Having multiple diagnostics on a span would cause the code fix
to throw an exception

1. How does that happen?  How would you have multiple variables marked as unused in the same span?  Each variable should be its own span right?
2. Can we have a test that shows what this fixes?

**Comment by Therzok** [comment]
See associated github issue for details. Sadly, the report came in via Twitter, so there's nothing actionable.

**Comment by heejaechang** [comment]
@Therzok Engine will collect all diagnostics at same span with diagnosticId the fixer claims can fix and give in. up to diagnostic analyzer and fixer for the diagnostic analyzer how to proceed with it. fix should be careful since our design only separate things based on diagnostic id, and we have no restriction on what diagnostic id each diagnostic analyzer can produce.

**Comment by CyrusNajmabadi** [comment]
>  up to diagnostic analyzer and fixer for the diagnostic analyzer how to proceed with it. 

The overall point of my question in https://github.com/dotnet/roslyn/pull/30093#issuecomment-423613023 is "what is the *right* way to proceed"?  It's very much not clear, and we are wildly inconsistent in the IDE in terms of how things are handled.  This is problematic because it means every time we write a new feature, people don't know what they should be doing, and people reviewing the code don't have an idea either.

It would be very helpful to get a writeup explaining when this can occur, as well as an explanation of how it should properly be handled.  Thanks!

**Comment by CyrusNajmabadi** [comment]
@heejaechang This went into 15.9.x.  Was there approval for that to happen?  What's the current bar/approval-process for that branch?  Thanks!

**Comment by Therzok** [comment]
Just in case, workitem 685771 for reference of bug on VSMac side.

**Comment by heejaechang** [comment]
oh. ya. didn't know this is targetting dev15.9.x, why is it targetting dev15.9? anyway, tagging @jinujoseph

**Comment by heejaechang** [comment]
@Therzok can you finish up ask mode template?

**Comment by heejaechang** [comment]
there is no right way, it is up to diagnostic fixer author how they want to proceed. getting Single is obviously wrong since engine never guarantee that there will be only 1 diagnostic given when fixer is called. but after that, fixer is fine to take the first one or second one or they choose to offer different fix based on whether there is 1 diagnostics, 2 diagnostics or more.

**Comment by CyrusNajmabadi** [comment]
> it is up to diagnostic fixer author how they want to proceed.

My point is: how is the fixer to know what is the best way to proceed?  Should they report a single fix per diagnostic?  Should they report one fix for *all* the diagnostics?  Should they check the first diagnostic?  Shoudl they check all diagnostics?

I ask because no one on the IDE team actually knows what the right thing is to do, which is why every feature is written differently.  There's no way to audit or review this stuff either because no one knows what the appropriate approach is if/when this happens.  

This is why i asked hte questoins in: https://github.com/dotnet/roslyn/pull/30093#issuecomment-423613023

Saying it's up to the author on how to proceed is unhelpful. I'm the author of about 90% of these, and i don't know what the right way is to proceed.  I'm tyring to ask the people who created this API how to go about understanding this area so i can:

1. understand the right way forward.
2. go audit all the existing fixers to do things properly.
3. properly review all future PRs in this space.

Right now i can't do any of those three (and pretty much most people making a PR or maintaining this code can't do it either).  That's not a good state to be in.  

--

Note: this is similar to the request i have given in teh past to document the IDiagnosticService.  These services and components are consumed by the IDE, but there are important gaps in understanding how best to use them properly.  When explanations are requested, it would be **super** helpful to keep that in mind as stating that it's just up to us how we want to handle it doesn't really solve anything, since we rose the question in the first place because we don't understand how to properly handle it.

**Comment by CyrusNajmabadi** [comment]
> oh. ya. didn't know this is targetting dev15.9.x, why is it targetting dev15.9? anyway, tagging @jinujoseph

Until this gets approval to have been merged, we should probably revert the merge.

**Comment by heejaechang** [comment]
> Should they report a single fix per diagnostic? Should they report one fix for all the diagnostics? Should they check the first diagnostic? Shoudl they check all diagnostics?

it is up to fixer like I said above. we don't have any contract on it. it can offer or not offer as many fix as they want for diagnostics they got and they are fine on how they interpret diagnostics given in.

**Comment by heejaechang** [comment]
> I ask because no one on the IDE team actually knows what the right thing is to do, which is why every feature is written differently. There's no way to audit or review this stuff either because no one knows what the appropriate approach is if/when this happens.

because there is no 1 way that is right. if one wants to keep behavior same regardless of what engine allows, sure they can create their own rule internally and keep it. but engine wise, it doesn't care.

also, if one wants to make those things consistent, create roslyn repo only analyzer to report inconsistency, what's one of use case we told poeple when we created analyzer.

**Comment by CyrusNajmabadi** [comment]
@heejaechang As an IDE dev, i'm asking you to help understand this space better.  I'm not sure how i can make it clearer.  The API is not understood well enough to properly use by the people consuming it (i.e. me and others who have tried to write things, and only have the existing analyzers to look at to go off of).

Telling me that it's up to me is not helpful because i don't know what the right thing is to do.  And saying you ca'nt say because it's up to me to figure out what's right is also not helpful *because i don't understand the semantics of the API*.

i.e. i don't understand what the distinction is between reporting a fix for one of the diagnostics or reporting a fix for *all* fo the diagnostics.

What is the *distinction*?  How do they differ?  What impact does it have in the engine wrt to all the analyzers/fixers it is running?  What impact does it have on the end user experience?

**Comment by heejaechang** [comment]
we are talking about this API
> http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/CodeFixes/CodeFixContext.cs,44

        /// <summary>
        /// Diagnostics to fix.
        /// NOTE: All the diagnostics in this collection have the same <see cref="CodeFixContext.Span"/>.
        /// </summary>

/// <summary>
        /// Creates a code fix context to be passed into <see cref="CodeFixProvider.RegisterCodeFixesAsync(CodeFixContext)"/> method.
        /// </summary>
        /// <param name="document">Document to fix.</param>
        /// <param name="span">Text span within the <paramref name="document"/> to fix.</param>
        /// <param name="diagnostics">
        /// Diagnostics to fix.
        /// All the diagnostics must have the same <paramref name="span"/>.
        /// Additionally, the <see cref="Diagnostic.Id"/> of each diagnostic must be in the set of the <see cref="CodeFixProvider.FixableDiagnosticIds"/> of the associated <see cref="CodeFixProvider"/>.
        /// </param>
        /// <param name="registerCodeFix">Delegate to register a <see cref="CodeAction"/> fixing a subset of diagnostics.</param>
        /// <param name="cancellationToken">Cancellation token.</param>
        /// <exception cref="ArgumentNullException">Throws this exception if any of the arguments is null.</exception>
        /// <exception cref="ArgumentException">
        /// Throws this exception if the given <paramref name="diagnostics"/> is empty,
        /// has a null element or has an element whose span is not equal to <paramref name="span"/>.
        /// </exception>
        public CodeFixContext(
            Document document,
            TextSpan span,
            ImmutableArray<Diagnostic> diagnostics,
            Action<CodeAction, ImmutableArray<Diagnostic>> registerCodeFix,
            CancellationToken cancellationToken)
            : this(document, span, diagnostics, registerCodeFix, verifyArguments: true, cancellationToken: cancellationToken)
        {
        }

so, where in the contract says the Diagnostics will have only 1?

,,

again, I dont think I can make you understand, so give me what you are understanding then I will point what assumption is wrong and then, after everyone happy, I can put that as doc comment to the API.

**Comment by heejaechang** [comment]
@CyrusNajmabadi so. I offered a way to fix it. tell me what is your understanding. and I will tell you what assumption you have is wrong and then when we agree there is no confusion, I will put that as doc comment for the API.

**Comment by CyrusNajmabadi** [comment]
I'm not making assumptions.  That's my point.  I literally do not understand how to effectively use this API because i do *not* know hte impact of the decisions made here.  

For example, i do not understand the impact that happens if you register a single fix versus many diagnostics.  What will happen next?  How does the engine use that information?  Will it cease to call other fixers?  What happens if it is multiple fixes?

WHat are the situations where you get multiple diagnostics?  What relationship do those diagnostics have to each other.

**Comment by CyrusNajmabadi** [comment]
Note: the reason a lot of the features were written this way was that a long time ago i explicitly asked: "wait... why/when does thsi get a set of diagnostics back?"

And i was told *explicitly* that that would not happen in practice, and it was borne out of an earlier design for analyzers where it was thought that this was how fix-all might work.  I believe that information came from Manish.  So it's very tough to know what is the right thing to do because the information from the past no longer seems correct, and there is very little data to indicate when/how this can happen, and what the potential impact is of the varying different approaches that can be taken.

**Comment by heejaechang** [comment]
someone wrote the code. so the one who wrote the code should have some kind of assumption when they wrote the code. otherwise, I don't get how they wrote the code.

again, as I said, you asked, so I gave you the doc comment and what to expect. and you tell me you have no idea how it works even though I tried to give you what you asked several times. so no point I keep trying since it seems you can't understand my words to your question.

either someone else answers your question, or give me what you think what I am saying then I will point what's wrong or confusion is and then we can go from there. or I don't see any point keep doing this.
I just want to make sure that I at least tried.

**Comment by CyrusNajmabadi** [comment]
> someone wrote the code.

I wrote the code. :)  In about 90% of the scenarios.

> so the one who wrote the code should have some kind of assumption when they wrote the code. otherwise, I don't get how they wrote the code.

The assumption was: i don't know how this works.  But this seems to work.  Talking to Manish in the past gave me the impression this was ok.

> and you tell me you have no idea how it works even though I tried to give you what you asked several times. 

 I literally still do not have answers for the very specific questions i've given.  The documentation does not cover those questions.  It feels you're deliberately ignoring those questions.

> either someone else answers your question, or give me what you think what I am saying then I will point what's wrong or confusion is and then we can go from there. or I don't see any point keep doing this.

Ok.  Once again (this is like the 4th or 5th time i'm asking this):

> For example, i do not understand the impact that happens if you register a single fix versus many diagnostics. What will happen next? How does the engine use that information? Will it cease to call other fixers? What happens if it is multiple fixes?

> WHat are the situations where you get multiple diagnostics? What relationship do those diagnostics have to each other.

> i.e. i don't understand what the distinction is between reporting a fix for one of the diagnostics or reporting a fix for all fo the diagnostics.

> What is the distinction? How do they differ? What impact does it have in the engine wrt to all the analyzers/fixers it is running? What impact does it have on the end user experience?

I cannot make an informed choice on how to use this API because i simply *do not now how using the API in these different ways actually affects the user experience*.  I do not know how to make it any simpler to ask this.  

To repeat: there is an api that accepts several different forms of input from the client (i.e. me).  I do not  know the impact of the different forms of data i can pass in.  So i do not know the right way to use the API.

**Comment by heejaechang** [comment]
>  i simply do not now how using the API in these different ways actually affects the user experience

I guess we are talking 2 different thing because the question is too broad. give more specific question then.

**Comment by CyrusNajmabadi** [comment]
I do not understand what is broad about these questions:

> i.e. i don't understand what the distinction is between reporting a fix for one of the diagnostics or reporting a fix for all fo the diagnostics.

>  What is the distinction? How do they differ? What impact does it have in the engine wrt to all the analyzers/fixers it is running? What impact does it have on the end user experience?

I don't know how to make it any simpler...

**Comment by CyrusNajmabadi** [comment]
ok.  let me try to make this simpler with an actual *real* example from code that is being reviewed *right now* (and which is a pattern we see all over the place).

What is the difference between:

```c#
        public override Task RegisterCodeFixesAsync(CodeFixContext context)
        {
            var document = context.Document;
            var diagnostic = context.Diagnostics[0];
            
            context.RegisterCodeFix(new MyCodeAction(
                c => FixAsync(document, diagnostic, c)),
                context.Diagnostics);

            return Task.CompletedTask;
        }

// and

        public override Task RegisterCodeFixesAsync(CodeFixContext context)
        {
            var document = context.Document;
            var diagnostic = context.Diagnostics[0];
            
            context.RegisterCodeFix(new MyCodeAction(
                c => FixAsync(document, diagnostic, c)),
                diagnostic);

            return Task.CompletedTask;
        }
```

?

What are potential pitfalls here.  What internal or external behavioral difference will this cause?

**Comment by heejaechang** [comment]
the API says. context.Diagnostics will return diagnostics at the same span as context.Span with diagnostics indicated by CodeFixProvider.FixableDiagnosticIds

so from the contract. that code above can have issue such as

1. it doesn't check diagnostic Id (CodeFixeProvider.FixableDiagnosticIds might return just 1 so that's why it might not checking, but if it returns multiple, then it should check)
2. it seems to assume that diagnostic analyzer that produces the diagnostic is one from it knows, but that's not true. code fixer never guaranteed that same diagnostic ID can be produced by only the same analyzer. a random analyzer from a third party can produce the same diagnostic id, and that will give given to this fixer if the span is the same. analyzer and fixer was loosely coupled from beginning and it was never changed since the very first initial design from compiler design meeting.

basically, the above code looks like having more restriction assumption on the input then the engine never promoised.

**Comment by heejaechang** [comment]
by the way, fix doesn't need to do checking diagnostic id. that's not only way to handle it. it can just try to see token at the span of the diagnostic. and see whether it is right one or not regardless of diagnostic ids.

when last time I checked most of IDE fixer did that. might have changed. but that was how things were when I saw one last time.

**Comment by CyrusNajmabadi** [comment]
That doesn't answer my question.

> it doesn't check diagnostic Id (CodeFixeProvider.FixableDiagnosticIds might return just 1 so that's why it might not checking, but if it returns multiple, then it should check)

Why would it need to check diagnostic IDs?  

--

You also didn't answer my explicit question in the post:

>  What internal or external behavioral difference will this cause?


I read your above post, and i still don't know what the right thing to do is here.  It is not good that that's the case.

**Comment by CyrusNajmabadi** [comment]
> when last time I checked most of IDE fixer did that. might have changed. but that was how things were when I saw one last time.

I did a spotcheck.  About 50+ cases where we do not do that.  The general view is htat "IDE" diagnostics are not a general extension point.  We only support them being produced/consumed by IDE code, with no other contract implicitly or explicitly given.  Consumption takes a hard dependency on the consumption code (which is why we keep them paired and we update them in lockstep).

**Comment by heejaechang** [comment]
> Why would it need to check diagnostic IDs?

because a fixer can say I can fix DiagnosticID1, ID2, and ID3 and provide completely different fix based on ID

> What internal or external behavioral difference will this cause?

can you be more specific. I am not following. are you asking what behavior change engine get by taking first diagnostic from dignsotics or [0] or single?

it used to have no impact, but IDE team has changed LB (icon and ordering) behavior based on severity. that probably you know better than me.

**Comment by CyrusNajmabadi** [comment]
> I keep trying to answer. and if it is not the form you exactly looking for, you just ignore. 

I literally asked as explicitly as i could manage.  I'm trying to ask very precise question looking for very specific information *that i have outlined explicitly*.  Yes, i am looking for that very specific information.  That's why i'm asking for it!

I asked Manish, because he provided useful information here in the past.  You've since started tryign to answer, but you haven't actually addressed the core *specific* questions that i laid out.  You haven't helped me understand the specific things i was asking for with *specific* code exapmles to demonstrate exactly what i was asking about.  

I literally do not know what i can do here.  I've tried to boil this down as simply as possible. And i've tried to make it clear *exactly* what information i need.  I don't know what else i can do...

**Comment by heejaechang** [comment]
sure. since that change has done by IDE team. we can bring in people who did. Ravi or Balaji did? or we can look history for it if we are curious.

also, before the change, it had no impact. so whathever it returns it doesn't have impact. but I am not sure when IDE changed the LB behavior they also audit the code and fix them up with new contract/expectation.

**Comment by CyrusNajmabadi** [comment]
This is going in circles.  I'm done.  This conversation has been completely unhelpful.  @mavasani If you can shed some light here, i would *really* appreciate it.  There is definitely a lot of confusion in terms of how best to use this API.  I've outlined my questions, and i've included some code examples to show where there is confusion.  Any illumination you can provide here would be *very* appreciated.

**Comment by heejaechang** [comment]
no about the RegisterFix API contract
http://source.roslyn.io/#Microsoft.CodeAnalysis.Workspaces/CodeFixes/CodeFixContext.cs,185

```csharp
/// <summary>
        /// Add supplied <paramref name="action"/> to the list of fixes that will be offered to the user.
        /// </summary>
        /// <param name="action">The <see cref="CodeAction"/> that will be invoked to apply the fix.</param>
        /// <param name="diagnostics">The subset of <see cref="Diagnostics"/> being addressed / fixed by the <paramref name="action"/>.</param>
```
I think it clearly says what it expects for the diagnostics.

among all diagnostics it got, it can return subset as it is fixing them. or return n different fixes with diagnostic it is fixing. it can be different diagnosticIDs or same diagnosticID but different form of fixes and etc.

**Comment by heejaechang** [comment]
well. you could work with me to get to the answer rather than just blaming and upset that I didn't give you what you asked me to provide. I am not here to serve you.

**Comment by CyrusNajmabadi** [comment]
> I think it clearly says what it expects for the diagnostics.

It literally says nothing that answers the question i have.  Specifically, what the impact/behavior change would be.

> well. you could work with me to get to the answer rather than just blaming and upset that I didn't give you what you asked me to provide. 

I have done everything in my power.  I have asked the questions multiple times in multiple ways. I provided you with code examples.  I asked clearly and specifically what i was looking for.  I literally do not know how i can communicate this any more clearly.

> I am not here to serve you

You chose to respond here.  I've taken a lot of time to try to converse here and make it possible to come to some conclusion. But you have steadfastly and consistently decided to not answer my actual questions.

You're right.  You don't serve me.  But you chose to try to answer the questions i laid out for Manish and since then you've posted several dozen responses that continue to work to not actually answer the questions.   This has been an enormous time suck that did not need to happen if you had jsut spent hte time to please read and respond to the questions i've asked.

That's why i'm done.  There is little point continuing this with you since you will simply keep doing this. I don't know if its intentional, or if it's a communication issue.  At this point, i've experienced so many times with some different PRs that i'm simply uninterested in engaging further.  No one else on the team behaves this way, and everyone else manages to provide useful and workable information with a bare minimum of contention or confusion.  So, frankly, at this point, i would just rather discuss things with others since these roundabout discussions *never accomplish anything* and you just seem to keep doing them with me over and over and over again.

**Comment by heejaechang** [comment]
ya. I wrote on many PRs as well. and you are the only one who responds to my comment this way as well. no idea what is upsetting you.

**Comment by jinujoseph** [comment]
@heejaechang this has not been approved for 15.9 and no approval tags associated too.
Pls revert and target the change to master

**Comment by heejaechang** [comment]
@Therzok can you target this to master? then I can check this in. tagging @jinujoseph in case @Therzok really want it to be in 15.9

**Comment by jinujoseph** [comment]
15.9 bar is very high bar at this point.
If not top issue, may not meet the bar.
Let me know if this is critical for 15.9

**Comment by Therzok** [comment]
Ask mode template filled.

**Comment by CyrusNajmabadi** [comment]
@mavasani Great.  Thanks.  Taht helped clear up a lot.  One area where i think it would be good to come to a general team rule on is:

> Additionally, it is recommended that the code fixer has defensive checks 

Right now, i believe 100% (or near 100%) of all IDE fixers don't do the defensive checks for any IDE00XX fixes.  The general feeling being that it just simply is not supported at all for anyone else ot be making these errors, and if you do it, and get it wrong, that's on you :)

If the team decides that shoudl change, we'd need to go fixup all the fixers.  However, my preference would be to jsut say "doing that is unsupported, and the expectation is IDE fixers can depend on only their own analyzers producing those diagnostics.

**Comment by mavasani** [comment]
> The general feeling being that it just simply is not supported at all for anyone else ot be making these errors, and if you do it, and get it wrong, that's on you :)

Yes, given that is a very rare case where you have a different analyzer package implementing IDE00XX, we can assume this is very unlikely scenario, which can be dealt with if and when someone reports it. However, such conservative checks are strongly recommended for other third party analyzer packages as diagnostic ID clashes are highly likely for them.
